README.md
# Generic API wrapper

version=0.2.0

`generic-api-wrapper` is a generic REST API python wrapper. It comes with an AWS IAM authenticator out of the box. 

# Getting started

## Install
`pip install generic-api-wrapper`

## Usage

### Basic usage
```python
from generic_api_wrapper import Api


# To create an API wrapper object provide base url
# Let's take Github for example
api = Api("https://api.github.com")

# Get repos of a user
# This is equivalent to sending a GET request to
# https://api.github.com/users/grishasergii/repos
repos = api.users.grishasergii.repos()

# If a resource name violates Python attributes syntax then use __getitem__
# This is equivalent to sending a GET request to
# https://api.i-do-not-exist.com/resources/imaginary-resource
imaginary_api = Api("https://api.i-do-not-exist.com")
resource = api.resources["imaginary-resource"]()

# provide data for POST request
imaginary_api.resources.post(data={"name": "resource-name"})
```

### Authentication
- AWS IAM authnetication
```python
from generic_api_wrapper import Api, AwsIamAuth


api_url = "https://api.aws-powered-api.com"

# AWS IAM authentication. Credentials and AWS region are inferred by boto3 from the app environment
auth = AwsIamAuth(api_url).auth

# Create an API wrapper object that is authenticated to call an API configured with an AWS IAM authorizer
api = Api(api_url, auth)
```
- custom authenticator: create a custom authenticator that fits your requirements 
and provide it to the `Api` constructor, it will be passed to the `session` object when sending a request.

# License

Free software: MIT license

# Credits

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

- Cookiecutter: https://github.com/audreyr/cookiecutter
- `audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.1.0 (2022-02-19)
------------------

* First release on PyPI.


tests/__init__.py
"""Unit test package for generic_api_wrapper."""


tests/conftest.py
import time

import docker
import pytest


@pytest.fixture(autouse=True)
def testing_env(monkeypatch):
    """Set ENV variables for testing."""
    monkeypatch.setenv("API_SERVER_PORT", "5002")
    monkeypatch.setenv("API_SERVER_URL", "http://127.0.0.1:5002")


@pytest.fixture(scope="module")
def api_server():
    """Starts a mock api server in a docker container."""
    image_name = "generic-api-wrapper-test-api-server-fixture"
    client = docker.from_env()
    client.images.build(path="tests/api_server", tag=image_name)
    container = client.containers.run(
        image=image_name,
        auto_remove=True,
        detach=True,
        ports={5000: 5002},
        name=image_name,
    )

    time.sleep(2)

    yield

    container.kill()


tests/test_wrapy.py
import os

import pytest
from requests.exceptions import HTTPError

from wrapy.wrapy import Api


@pytest.mark.parametrize("url", ["https://api.github.com", "https://api.github.com/"])
def test_github_api_existing_path_returns_valid_data(url):
    github = Api(url)
    repos = github.users.grishasergii.repos()
    assert repos


def test_github_api_path_does_not_exist_returns_not_found():
    github = Api("https://api.github.com")
    with pytest.raises(HTTPError):
        response = github.users.agvekgvace757362fgvj23yt.repos()
        pass


def test_getitem_and_getattr_returns_valid_response(api_server):
    api = Api(os.environ["API_SERVER_URL"])
    resource_id = "test-resource-1"
    response = api.tests[resource_id]()

    assert response["data"] == resource_id


wrapy/wrapy.py
"""Main module."""
import requests
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry


class RetrySession:  # pylint: disable=too-few-public-methods
    """Session object with retry capabilities."""

    def __init__(self):
        self.session = None

    def _requests_retry_session(
        self, retries=5, backoff_factor=2, status_forcelist=(500, 502, 503, 504)
    ):
        """Returns a retriable session."""
        if self.session:
            return self.session

        session = requests.Session()
        retry = Retry(
            total=retries,
            read=retries,
            connect=retries,
            backoff_factor=backoff_factor,
            status_forcelist=status_forcelist,
            allowed_methods={"GET"},
        )
        adapter = HTTPAdapter(max_retries=retry)
        session.mount("https://", adapter)

        self.session = session

        return self.session

    def get(self, url, **kwargs):
        """GET method forwarded to a session object."""
        return self._requests_retry_session().get(url, **kwargs)

    def post(self, url, **kwargs):
        """POST method forwarded to a session object."""
        return self._requests_retry_session().post(url, **kwargs)

    def delete(self, url, **kwargs):
        """DELETE method forwarded to a session object."""
        return self._requests_retry_session().delete(url, **kwargs)


class Api:
    """Api wrapper object."""

    def __init__(self, base_url, auth=None):
        """Creates an API wrapper object.

        Args:
            base_url: string
            auth: Auth tuple or callable to enable authentication, passed to the session
        """
        self.base_url = base_url.rstrip("/")
        self._session = RetrySession()
        self.parent = None
        self._auth = auth

    @classmethod
    def _from_parent(cls, url, parent):
        api = cls(url)
        api.parent = parent
        return api

    def __getattr__(self, item):
        new_base = f"{self.base_url}/{item}"
        return self.__class__._from_parent(new_base, self)

    def __getitem__(self, item):
        new_base = f"{self.base_url}/{item}"
        return self.__class__._from_parent(new_base, self)

    def __call__(self, **kwargs):
        return self.get(**kwargs)

    @property
    def session(self):
        if self.parent:
            return self.parent.session
        return self._session

    @property
    def auth(self):
        if self.parent:
            return self.parent.auth
        return self._auth

    def _make_request(self, method, **kwargs):
        session_method = getattr(self.session, method)
        response = session_method(self.base_url + "/", auth=self.auth, **kwargs)

        response.raise_for_status()
        if response.text:
            return response.json()
        return {}

    def get(self, **kwargs):
        """GET request."""
        return self._make_request("get", **kwargs)

    def post(self, **kwargs):
        """POST request."""
        return self._make_request("post", **kwargs)

    def delete(self, **kwargs):
        """DELETE request."""
        return self._make_request("delete", **kwargs)


generic_api_wrapper/generic_api_wrapper.py
"""Main module."""
import requests
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry


class RetrySession:  # pylint: disable=too-few-public-methods
    """Session object with retry capabilities."""

    def __init__(self):
        self.session = None

    def _requests_retry_session(
        self, retries=5, backoff_factor=2, status_forcelist=(500, 502, 503, 504)
    ):
        """Returns a retriable session."""
        if self.session:
            return self.session

        session = requests.Session()
        retry = Retry(
            total=retries,
            read=retries,
            connect=retries,
            backoff_factor=backoff_factor,
            status_forcelist=status_forcelist,
            allowed_methods={"GET"},
        )
        adapter = HTTPAdapter(max_retries=retry)
        session.mount("https://", adapter)

        self.session = session

        return self.session

    def get(self, url, **kwargs):
        """GET method forwarded to a session object."""
        return self._requests_retry_session().get(url, **kwargs)

    def post(self, url, **kwargs):
        """POST method forwarded to a session object."""
        return self._requests_retry_session().post(url, **kwargs)

    def delete(self, url, **kwargs):
        """DELETE method forwarded to a session object."""
        return self._requests_retry_session().delete(url, **kwargs)


class Api:
    """Api wrapper object."""

    def __init__(self, base_url, auth=None):
        """Creates an API wrapper object.

        Args:
            base_url: string
            auth: Auth tuple or callable to enable authentication, passed to the session
        """
        self.base_url = base_url.rstrip("/")
        self._session = RetrySession()
        self.parent = None
        self._auth = auth

    @classmethod
    def _from_parent(cls, url, parent):
        api = cls(url)
        api.parent = parent
        return api

    def __getattr__(self, item):
        new_base = f"{self.base_url}/{item}"
        return self.__class__._from_parent(new_base, self)

    def __getitem__(self, item):
        new_base = f"{self.base_url}/{item}"
        return self.__class__._from_parent(new_base, self)

    def __call__(self, **kwargs):
        return self.get(**kwargs)

    @property
    def session(self):
        if self.parent:
            return self.parent.session
        return self._session

    @property
    def auth(self):
        if self.parent:
            return self.parent.auth
        return self._auth

    def _make_request(self, method, **kwargs):
        session_method = getattr(self.session, method)
        response = session_method(self.base_url, auth=self.auth, **kwargs)

        response.raise_for_status()
        if response.text:
            return response.json()
        return {}

    def get(self, **kwargs):
        """GET request."""
        return self._make_request("get", **kwargs)

    def post(self, **kwargs):
        """POST request."""
        return self._make_request("post", **kwargs)

    def put(self, **kwargs):
        """PUT request."""
        return self._make_request("post", **kwargs)

    def delete(self, **kwargs):
        """DELETE request."""
        return self._make_request("delete", **kwargs)
